[ValueTracking] Add `matchSimpleBinaryIntrinsicRecurrence` helper

antoniofrighetto opened the pull request:
Similarly to what it is being done to match simple recurrence cycle relations, attempt to match value-accumulating recurrences of kind:
```
  %umax.acc = phi i8 [ %umax, %backedge ], [ %a, %entry ]
  %umax = call i8 @llvm.umax.i8(i8 %umax.acc, i8 %b)
```
Preliminary work to let InstCombine avoid folding such recurrences, so that simple loop-invariant computation may get hoisted. Minor opportunity to refactor out code as well.

dtcxzyw commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -9119,6 +9121,23 @@ bool llvm::matchSimpleRecurrence(const BinaryOperator *I, PHINode *&P,
+  if (!P)
+    P = dyn_cast<PHINode>(I->getArgOperand(1));
+
+  std::function<bool(const IntrinsicInst *)> IsBinaryOrMinMaxFn =
```
Can we run the predicate after matching the recurrence?

antoniofrighetto replied:
We could run the predicate after matching the recurrence, but that would imply computing all the recurrence needlessly, if the intrinsic turned out not to be a candidate one (as we're passing a general IntrinsicInst), in addition to checking intrinsic operands size (as not to go out of bound when retrieving the Operation ops). Switched this to using a lightweight llvm::function_ref instead, I think it should be quite better.

dtcxzyw commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -9119,6 +9121,23 @@ bool llvm::matchSimpleRecurrence(const BinaryOperator *I, PHINode *&P,
+
+  std::function<bool(const IntrinsicInst *)> IsBinaryOrMinMaxFn =
+      [](const IntrinsicInst *I) {
+        return isa<BinaryOpIntrinsic, MinMaxIntrinsic>(I);
```
`xx.with.overflow` should not be accepted. BTW, I'd like to handle fmax/fmin as well.

```suggestion
        return isa<SaturatingInst, MinMaxIntrinsic>(I);
```

dtcxzyw reviewed (commented):
Can you add some unit tests to `llvm/unittests/Analysis/ValueTrackingTest.cpp`?

nikic commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -9119,6 +9120,28 @@ bool llvm::matchSimpleRecurrence(const BinaryOperator *I, PHINode *&P,
+                      IID == Intrinsic::minimum || IID == Intrinsic::maximum ||
+                      IID == Intrinsic::minimumnum ||
+                      IID == Intrinsic::maximumnum;
+             }) &&
```
I would expect that the caller is responsible for checking the intrinsic ID. We should only check that it's a two argument intrinsic (maybe with same arg and ret types) here.

nikic reviewed (approved):
LGTM

antoniofrighetto commented on llvm/include/llvm/Analysis/ValueTracking.h:
```diff
@@ -965,6 +966,21 @@ LLVM_ABI bool matchSimpleRecurrence(const PHINode *P, BinaryOperator *&BO,
 
+/// Attempt to match a simple value-accumulating recurrence of the form:
+///   %llvm.intrinsic.acc = phi Ty [%Init, %Entry], [%llvm.intrinsic, %backedge]
+///   %llvm.intrinsic = call Ty @llvm.intrinsic(%Invariant, %llvm.intrinsic.acc)
```
Right, was mentioned in the doc-comment, but the variable name didn't reflect this; updated, thanks!

antoniofrighetto committed:
[ValueTracking] Add `matchSimpleBinaryIntrinsicRecurrence` helper

Similarly to what it is being done to match simple recurrence cycle
relations, attempt to match value-accumulating recurrences of kind:
```
  %umax.acc = phi i8 [ %umax, %backedge ], [ %a, %entry ]
  %umax = call i8 @llvm.umax.i8(i8 %umax.acc, i8 %b)
```
Preliminary work to let InstCombine avoid folding such recurrences,
so that simple loop-invariant computation may get hoisted. Minor
opportunity to refactor out code as well.

antoniofrighetto commented on llvm/include/llvm/Analysis/ValueTracking.h:
```diff
@@ -21,6 +21,7 @@
 #include "llvm/IR/FMF.h"
 #include "llvm/IR/InstrTypes.h"
 #include "llvm/IR/Instructions.h"
+#include "llvm/IR/IntrinsicInst.h"
```
Fixed in 69b8e59f29a38a6bdfa0d506dae8d8deccf42862.
